blog/tanstack-router-loaders/index.md

```diff
@@ -0,0 +1,630 @@
+
+#### beforeLoad
+
+`beforeLoad` runs always, on each active route, anytime the url changes in any way. This is a good place to check preconditions and redirect. If you return a value from here, that value will be merged into the router's context, and visible from that route, downward. This function **blocks** all loaders from running, so be **extremely careful** what you do in here. Data loading should generally be avoided unless absolutely needed, since any loaders will wait until this function is complete, potentially creating waterfalls.
+
+Here's a good example of what to avoid, with an opportunity to see why. This `beforeLoad` fetches the current user, and places it into context, and doing a redirect if there is no user.
+
+```ts
+  async beforeLoad({}) {
+    const user = await getCurrentUser();
+    if (!user) {
+      throw redirect({
+        to: "/login",
+      });
+    }
+    document.cookie = `user=${user.id};path=/;max-age=31536000`;
+
+    return { user };
+  },
+```
+
+We'll be looking at some data loading in a bit, and measure what starts when. You can go into the `getCurrentUser` function and uncomment the artificial delay in there, and see it block _everything_. This is especially obvious if you're running Router's dev tools. You'll see this path block, and only once ready, allow all loaders below to execute.
+
+But this is a good enough example to show how this works. The `user` object is now in context, visible to routes beneath it.
+
+**NOTE**
+
+```
+A more realistic example would probably *just* check for a logged-in cookie, and optimistically assume the user is logged in, and rely on network calls we do in the loaders to detect a logged-out user, and redirect accordingly. And to make things even more realistic, those loaders for the initial render would run on the server, and figure out if a user is actually logged out before we show the user *anything*; but that will wait for a future post on TanStack Start.
+
+What we have is sufficient to show how the `beforeLoad` callback works.
+```
+
+#### Context (function)
+
+There's also a context `function` we can provide routes. This is a non-async function that also gives us an opportunity to add to context. But it runs much more conservatively. This function only runs when the url changes in a way that's relevant to _that route_. So for a route of, say, `app/epics/$epicId`, the context function will re-run when the epicId param changes. This might seem strange, but it's useful for modifying the context, but only when the route has changed, especially when you need to put non-primitive values (objects and functions) onto context. These non-primitive values are always compared by reference, and therefore always unique against the last value generated. As a result, they would cause render churning if added in `beforeLoad`, since Router would (incorrectly) think it needed to re-render a route when nothing has changed.
```

Comment:
> As a result, they would cause render churning if added in `beforeLoad`, since Router would (incorrectly) think it needed to re-render a route when nothing has changed.

Router is not doing any re-rendering when refs change. That's React's job!

Reply:
done